Build failure: rocmPackages.llvm.libc

## Steps To Reproduce

Steps to reproduce the behavior:

nix build github:NixOS/nixpkgs#rocmPackages.llvm.libc

## Build log

[log.txt](https://github.com/user-attachments/files/18264262/log.txt)

<details>
<summary>Short version</summary>

```
error: builder for '/nix/store/axshsp2rv451cn9cqhn0djrcjdkr9l0b-rocm-llvm-libc-6.0.2.drv' failed with exit code 1;
       last 25 log lines:
       > #include <cassert>
       >          ^~~~~~~~~
       > 1 error generated.
       > [4/322] Building CXX object libc/utils/testutils/CMakeFiles/libc_test_utils.dir/Timer.cpp.o
       > FAILED: libc/utils/testutils/CMakeFiles/libc_test_utils.dir/Timer.cpp.o
       > /nix/store/7sqn5142gh5asri8aqxvsvib9ii37fyq-rocm-llvm-clang-unwrapped-wrapper-6.0.2/bin/clang++ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=gnu++17 -MD -MT libc/utils/testutils/CMakeFiles/libc_test_utils.dir/Timer.cpp.o -MF libc/utils/testutils/CMakeFiles/libc_test_utils.dir/Timer.cpp.o.d -o libc/utils/testutils/CMakeFiles/libc_test_utils.dir/Timer.cpp.o -c /build/source/libc/utils/testutils/Timer.cpp
       > /build/source/libc/utils/testutils/Timer.cpp:11:10: fatal error: 'chrono' file not found
       > #include <chrono>
       >          ^~~~~~~~
       > 1 error generated.
       > [5/322] Building CXX object libc/src/errno/CMakeFiles/libc.src.errno.errno.dir/libc_errno.cpp.o
       > [6/322] Building CXX object libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o
       > FAILED: libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o
       > /nix/store/7sqn5142gh5asri8aqxvsvib9ii37fyq-rocm-llvm-clang-unwrapped-wrapper-6.0.2/bin/clang++ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=gnu++17 -MD -MT libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o -MF libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o.d -o libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o -c /build/source/libc/utils/testutils/ExecuteFunctionUnix.cpp
       > /build/source/libc/utils/testutils/ExecuteFunctionUnix.cpp:10:10: fatal error: 'cassert' file not found
       > #include <cassert>
       >          ^~~~~~~~~
       > 1 error generated.
       > [7/322] Building CXX object libc/utils/testutils/CMakeFiles/libc_test_utils.dir/RoundingModeUtils.cpp.o
       > /build/source/libc/utils/testutils/RoundingModeUtils.cpp:30:3: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
       >   default:
       >   ^
       > 1 warning generated.
       > [8/322] Building CXX object libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o
       > ninja: build stopped: subcommand failed.

```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.66-hardened1, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - nixpkgs: `/nix/store/8fwsiv0hd7nw1brkvka0jf1frk3m7qkr-source`


## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@acowley 
@lovesegfault 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

With:
```
nix-build --arg overlays '[(self: super: { p = self.rocmPackages.llvm.libc.overrideAttrs (prev: { NIX_DEBUG = 1; } ); })]' /nix/store/8vz84mqgnm1gz5yk7hgnnb5gir5hjxas-source -A p
```

I was able to get NIX_DEBUG into the args and see that `-nostdlibinc` into the arguments. So the question is why it would be being inserted there.

```
/build/source/libc/test/UnitTest/LibcTest.cpp:16:10: fatal error: 'cassert' file not found
#include <cassert>
         ^~~~~~~~~
1 error generated.
[32/322] Building CXX object libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o
HARDENING: disabled flags: pacret pie stackclashprotection fortify3 shadowstack fortify trivialautovarinit
HARDENING: Is active (not completely disabled with "all" flag)
HARDENING: enabling pic
HARDENING: enabling format
HARDENING: enabling zerocallusedregs
HARDENING: enabling stackprotector
HARDENING: enabling strictoverflow
extra flags before to /nix/store/dcakkwyrmphjj0vqdi7h8bd570mnv6pr-rocm-llvm-clang-unwrapped-6.0.2/bin/clang++:
  -fPIC
  -Wformat
  -Wformat-security
  -Werror=format-security
  -fzero-call-used-regs=used-gpr
  -fstack-protector-strong
  --param
  ssp-buffer-size=4
  -fwrapv
  -target
  x86_64-unknown-linux-gnu
original flags to /nix/store/dcakkwyrmphjj0vqdi7h8bd570mnv6pr-rocm-llvm-clang-unwrapped-6.0.2/bin/clang++:
  -D__STDC_CONSTANT_MACROS
  -D__STDC_FORMAT_MACROS
  -D__STDC_LIMIT_MACROS
  -I/build/source/libc
  -fvisibility-inlines-hidden
  -Werror=date-time
  -Werror=unguarded-availability-new
  -Wall
  -Wextra
  -Wno-unused-parameter
  -Wwrite-strings
  -Wcast-qual
  -Wmissing-field-initializers
  -Wimplicit-fallthrough
  -Wcovered-switch-default
  -Wno-noexcept-type
  -Wnon-virtual-dtor
  -Wdelete-non-virtual-dtor
  -Wsuggest-override
  -Wstring-conversion
  -Wmisleading-indentation
  -Wctad-maybe-unsupported
  -fdiagnostics-color
  -ffunction-sections
  -fdata-sections
  -O3
  -DNDEBUG
  -std=gnu++17
  -MD
  -MT
  libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o
  -MF
  libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o.d
  -o
  libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o
  -c
  /build/source/libc/test/UnitTest/TestLogger.cpp
extra flags after to /nix/store/dcakkwyrmphjj0vqdi7h8bd570mnv6pr-rocm-llvm-clang-unwrapped-6.0.2/bin/clang++:
  -B/nix/store/65h17wjrrlsj2rj540igylrx7fqcd6vq-glibc-2.40-36/lib/
  -idirafter
  /nix/store/kj8hbqx4ds9qm9mq7hyikxyfwwg13kzj-glibc-2.40-36-dev/include
  --gcc-toolchain=/nix/store/70bgv05075sxjgpm5s31kvpq35iip0fa-gcc-12.4.0
  -B/nix/store/dcakkwyrmphjj0vqdi7h8bd570mnv6pr-rocm-llvm-clang-unwrapped-6.0.2/lib
  -nostdlibinc
  -resource-dir=/nix/store/7sqn5142gh5asri8aqxvsvib9ii37fyq-rocm-llvm-clang-unwrapped-wrapper-6.0.2/resource-root
  -fuse-ld=lld
  -rtlib=compiler-rt
  -unwindlib=libunwind
  -Wno-unused-command-line-argument
  -B/nix/store/f1746bc8sw3banh3pnxsqmrzjh1jjy6i-rocm-llvm-binutils-wrapper-6.0.2/bin/
  -frandom-seed=a18fns3m5f
  -isystem
  /nix/store/17gplrlgbkpxzviwqg988n9ajybjaw4c-python3-3.12.8-env/include
  -isystem
  /nix/store/c9m6yd8fg1flz2j5r4bif1ib5j20a0cy-python3-3.12.8/include
  -isystem
  /nix/store/8gq75qdv7yl0ci4icfyqi4nvgm47pcj4-rocm-llvm-runtimes-6.0.2/include
  -isystem
  /nix/store/129shslczvy20i9dg8986cvr4hal2zzz-rocm-llvm-llvm-6.0.2/include
  -isystem
  /nix/store/1lggwqzapn5mn49l9zy4h566ysv9kzdb-zlib-1.3.1-dev/include
  -isystem
  /nix/store/3lbb9lx1gmis6qm8clq9g9k37i5l8l52-ncurses-6.4.20221231-dev/include
  -isystem
  /nix/store/mkx2jc5qv43xni5a6zy7k2l862vjibzd-rocm-llvm-lld-6.0.2/include
  -isystem
  /nix/store/7hwyxbsi39mzrl4kmqs0m478wd2pg63x-libxml2-2.13.5-dev/include
  -isystem
  /nix/store/6cr0spsvymmrp1hj5n0kbaxw55w1lqyp-libxcrypt-4.4.36/include
  -isystem
  /nix/store/zrvw84q27y9dkly1d6980vn9j14b75lv-libedit-20240808-3.1-dev/include
  -isystem
  /nix/store/w661s7ai2rbh411r3w94fbfp802xnz8y-libffi-3.4.6-dev/include
  -isystem
  /nix/store/9hrap3ppvrdhivyazrvf04h4kqxlr9w2-mpfr-4.2.1-dev/include
  -isystem
  /nix/store/5p5arwim42ci73rkwr412p3ab3jj37ic-gmp-with-cxx-6.3.0-dev/include
  -isystem
  /nix/store/17gplrlgbkpxzviwqg988n9ajybjaw4c-python3-3.12.8-env/include
  -isystem
  /nix/store/c9m6yd8fg1flz2j5r4bif1ib5j20a0cy-python3-3.12.8/include
  -isystem
  /nix/store/8gq75qdv7yl0ci4icfyqi4nvgm47pcj4-rocm-llvm-runtimes-6.0.2/include
  -isystem
  /nix/store/129shslczvy20i9dg8986cvr4hal2zzz-rocm-llvm-llvm-6.0.2/include
  -isystem
  /nix/store/1lggwqzapn5mn49l9zy4h566ysv9kzdb-zlib-1.3.1-dev/include
  -isystem
  /nix/store/3lbb9lx1gmis6qm8clq9g9k37i5l8l52-ncurses-6.4.20221231-dev/include
  -isystem
  /nix/store/mkx2jc5qv43xni5a6zy7k2l862vjibzd-rocm-llvm-lld-6.0.2/include
  -isystem
  /nix/store/7hwyxbsi39mzrl4kmqs0m478wd2pg63x-libxml2-2.13.5-dev/include
  -isystem
  /nix/store/6cr0spsvymmrp1hj5n0kbaxw55w1lqyp-libxcrypt-4.4.36/include
  -isystem
  /nix/store/zrvw84q27y9dkly1d6980vn9j14b75lv-libedit-20240808-3.1-dev/include
  -isystem
  /nix/store/w661s7ai2rbh411r3w94fbfp802xnz8y-libffi-3.4.6-dev/include
  -isystem
  /nix/store/9hrap3ppvrdhivyazrvf04h4kqxlr9w2-mpfr-4.2.1-dev/include
  -isystem
  /nix/store/5p5arwim42ci73rkwr412p3ab3jj37ic-gmp-with-cxx-6.3.0-dev/include
```

There is probably something obvious wrong here but I don't have a lot of attention on this right now.

FYI this seems to affect rocmPackages_5 aswell

```
error: builder for '/nix/store/9l7mbqkfl1ip02zq0wkffwcpcmd7mcfz-rocm-llvm-libc-5.7.1.drv' failed with exit code 1;
       last 25 log lines:
       >   ^
       > 1 warning generated.
       > [7/322] Building CXX object libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o
       > FAILED: libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o
       > /nix/store/ms1026km7xhf0p64kifm1pxzgi0dkv2d-rocm-llvm-clang-unwrapped-wrapper-5.7.1/bin/clang++ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=gnu++17 -MD -MT libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o -MF libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o.d -o libc/utils/testutils/CMakeFiles/libc_test_utils.dir/ExecuteFunctionUnix.cpp.o -c /build/source/libc/utils/testutils/ExecuteFunctionUnix.cpp
       > /build/source/libc/utils/testutils/ExecuteFunctionUnix.cpp:10:10: fatal error: 'cassert' file not found
       > #include <cassert>
       >          ^~~~~~~~~
       > 1 error generated.
       > [8/322] Building CXX object libc/src/stdio/printf_core/CMakeFiles/libc.src.stdio.printf_core.writer.dir/writer.cpp.o
       > [9/322] Building CXX object libc/src/ctype/CMakeFiles/libc.src.ctype.isalnum.dir/isalnum.cpp.o
       > [10/322] Building CXX object libc/src/__support/CPP/CMakeFiles/libc.src.__support.CPP.new.dir/new.cpp.o
       > [11/322] Building CXX object libc/src/math/generic/CMakeFiles/libc.src.math.generic.math_utils.dir/math_utils.cpp.o
       > [12/322] Building CXX object libc/src/__support/StringUtil/CMakeFiles/libc.src.__support.StringUtil.signal_to_string.dir/signal_to_string.cpp.o
       > [13/322] Building CXX object libc/src/__support/StringUtil/CMakeFiles/libc.src.__support.StringUtil.error_to_string.dir/error_to_string.cpp.o
       > [14/322] Building CXX object libc/src/__support/File/CMakeFiles/libc.src.__support.File.file.dir/file.cpp.o
       > [15/322] Building CXX object libc/test/UnitTest/CMakeFiles/LibcUnitTest.dir/LibcTest.cpp.o
       > FAILED: libc/test/UnitTest/CMakeFiles/LibcUnitTest.dir/LibcTest.cpp.o
       > /nix/store/ms1026km7xhf0p64kifm1pxzgi0dkv2d-rocm-llvm-clang-unwrapped-wrapper-5.7.1/bin/clang++ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/build/source/libc -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=gnu++17 -MD -MT libc/test/UnitTest/CMakeFiles/LibcUnitTest.dir/LibcTest.cpp.o -MF libc/test/UnitTest/CMakeFiles/LibcUnitTest.dir/LibcTest.cpp.o.d -o libc/test/UnitTest/CMakeFiles/LibcUnitTest.dir/LibcTest.cpp.o -c /build/source/libc/test/UnitTest/LibcTest.cpp
       > /build/source/libc/test/UnitTest/LibcTest.cpp:16:10: fatal error: 'cassert' file not found
       > #include <cassert>
       >          ^~~~~~~~~
       > 1 error generated.
       > [16/322] Building CXX object libc/test/UnitTest/CMakeFiles/TestLogger.dir/TestLogger.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/9l7mbqkfl1ip02zq0wkffwcpcmd7mcfz-rocm-llvm-libc-5.7.1.drv'.
error (ignored): error: cannot unlink '"/tmp/nix-build-rocm-llvm-compiler-rt-5.7.1.drv-0/build/source"': Directory not empty
error: 1 dependencies of derivation '/nix/store/0i7p5jqxjz635i29gqi2csksln1db7yz-rocm-llvm-clang-wrapper-5.7.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/d9b8385h5ndxbh66j5b3ayhwqx6lw9z6-clr-5.7.1.drv' failed to build
```

--- a/pkgs/build-support/cc-wrapper/default.nix
+++ b/pkgs/build-support/cc-wrapper/default.nix
@@ -639,7 +639,7 @@ stdenvNoCC.mkDerivation {
     # no `/usr/include`, there’s essentially no risk to dropping
     # the flag there. See discussion in NixOS/nixpkgs#191152.
     #
-    + optionalString ((cc.isClang or false) && !targetPlatform.isDarwin) ''
+    + optionalString ((cc.isClang or false) && !(cc.isROCm or false) && !targetPlatform.isDarwin) ''
       echo " -nostdlibinc" >> $out/nix-support/cc-cflags
     ''
 

--- a/pkgs/development/rocm-modules/5/llvm/base.nix
+++ b/pkgs/development/rocm-modules/5/llvm/base.nix
@@ -180,6 +180,7 @@ stdenv.mkDerivation (finalAttrs: {
   passthru = {
     isLLVM = targetDir == "llvm";
     isClang = targetDir == "clang" || builtins.elem "clang" targetProjects;
+    isROCm = true;
 
     updateScript = rocmUpdateScript {
       name = finalAttrs.pname;

--- a/pkgs/development/rocm-modules/5/llvm/stage-3/clang.nix
+++ b/pkgs/development/rocm-modules/5/llvm/stage-3/clang.nix
@@ -40,7 +40,10 @@ wrapCCWith rec {
     '';
 
     passthru.isClang = true;
+    passthru.isROCm = true;
   });
+
+  gccForLibs = stdenv.cc.cc;
 
   extraPackages = [
     llvm

--- a/pkgs/development/rocm-modules/6/llvm/base.nix
+++ b/pkgs/development/rocm-modules/6/llvm/base.nix
@@ -188,6 +188,7 @@ stdenv.mkDerivation (finalAttrs: {
   passthru = {
     isLLVM = targetDir == "llvm";
     isClang = targetDir == "clang" || builtins.elem "clang" targetProjects;
+    isROCm = true;
 
     updateScript = rocmUpdateScript {
       name = finalAttrs.pname;

--- a/pkgs/development/rocm-modules/6/llvm/stage-3/clang.nix
+++ b/pkgs/development/rocm-modules/6/llvm/stage-3/clang.nix
@@ -40,7 +40,10 @@ wrapCCWith rec {
     '';
 
     passthru.isClang = true;
+    passthru.isROCm = true;
   });
+
+  gccForLibs = stdenv.cc.cc;
 
   extraPackages = [
     llvm

